File: Y2Q1/SDB/Project/Code/Scripts/pbfe2/src/main/scala/PBFE.scala
package main

import org.apache.spark.sql.{SQLContext, Point, Row}
import org.apache.spark.{SparkConf, SparkContext}

object PBFE {
  case class PointItem(id: Int, x: Double, y: Double)
  case class Pair(id1: Int, id2: Int, x1: Double, y1: Double, x2: Double, y2: Double)
  
  var epsilon: Double = 1.0
  var r2 = math.pow(epsilon/2,2)
  var X = 0.0
  var Y = 0.0
  var D2 = 0.0
  var root = 0.0
  var h1 = 0.0
  var h2 = 0.0
  var k1 = 0.0
  var k2 = 0.0
  
  def calculateDisks(pair: Row) : Pair = {
    X = pair.getDouble(1) - pair.getDouble(4)
    Y = pair.getDouble(2) - pair.getDouble(5)
    D2 = math.pow(X, 2) + math.pow(Y, 2)
    if (D2 == 0)
        null
    root = math.pow(math.abs(4.0 * (r2 / D2) - 1.0), 0.5)
    h1 = ((X + Y * root) / 2) + pair.getDouble(4)
    h2 = ((X - Y * root) / 2) + pair.getDouble(4)
    k1 = ((Y - X * root) / 2) + pair.getDouble(5)
    k2 = ((Y + X * root) / 2) + pair.getDouble(5)
    
    Pair(pair.getInt(0), pair.getInt(3), h1, k1, h2, k2)
  }
  
  def main(args: Array[String]) : Unit = {
    val sparkConf = new SparkConf().setAppName("PBFE").setMaster("local[*]")
    val sc = new SparkContext(sparkConf)
    val sqlContext = new SQLContext(sc)
    if(args.length == 2)
      sc.setLogLevel("ERROR")
    else
      sc.setLogLevel(args(2))
    sqlContext.setConf("spark.sql.shuffle.partitions", 4.toString)
    sqlContext.setConf("spark.sql.sampleRate", 1.toString)
    sqlContext.setConf("spark.sql.partitioner.strTransferThreshold", 1000000.toString)

    import sqlContext.implicits._

    val filename = args(0)
    epsilon = args(1).toDouble
    val tag = filename.substring(filename.lastIndexOf("/") + 1).split("\\.")(0).substring(1)

    val p1 = sc.textFile(filename).map(_.split(",")).map(p => PointItem(p(0).trim.toInt, p(1).trim.toDouble, p(2).trim.toDouble)).toDF()
    val p2 = p1.toDF("id2", "x2", "y2")

    val time1 = System.currentTimeMillis()
    val pairs = p1.distanceJoin(p2, Point(p1("x"), p1("y")), Point(p2("x2"), p2("y2")), epsilon)
    val disks = pairs.rdd
       .filter( (x:Row) => x.getInt(0) > x.getInt(3) )
       .map( (x: Row) => calculateDisks(x) )
    var n = disks.count()
    // pairs.collect()
    val time2 = System.currentTimeMillis()
    println("PBFE," + epsilon + "," + tag + "," + 2 * n + "," + (time2 - time1) / 1000.0)

    sc.stop()
  }
}

File: Y2Q1/SDB/Project/Code/Scripts/pbfe4/pbfe.py
#!/usr/bin/env python3
import math
import time
import sys
from os.path import basename, splitext
from pyspark import SparkConf, SparkContext
from pyspark.sql import SQLContext, Row

conf = (SparkConf().setMaster("local[*]").setAppName("PBFE"))
sc = SparkContext(conf = conf)
sqlContext = SQLContext(sc)

filename = sys.argv[1]
epsilon = float(sys.argv[2])
r2 = math.pow(epsilon/2,2)
tag = splitext(basename(filename))[0][1:]

def calculateDisks(pair):
    global epsilon
    global r2
    
    X = pair[1] - pair[4]
    Y = pair[2] - pair[5]
    D2 = math.pow(X, 2) + math.pow(Y, 2)
    if (D2 == 0):
        return []
    expression = abs(4 * (r2 / D2) - 1)
    root = math.pow(expression, 0.5)
    h1 = ((X + Y * root) / 2) + pair[4]
    h2 = ((X - Y * root) / 2) + pair[4]
    k1 = ((Y - X * root) / 2) + pair[5]
    k2 = ((Y + X * root) / 2) + pair[5]
    
    return Row(id1=pair[0],id2=pair[3],lat1=h1,lng1=k1,lat2=h2,lng2=k2)

points = sc.textFile(filename).map(lambda line: line.split(",")).map(lambda p: Row(id=p[0], lat=float(p[1]), lng=float(p[2]))).toDF()
t1 = time.time()
points.registerTempTable("p1")
points.registerTempTable("p2")
sql = """
    SELECT 
        * 
    FROM 
        p1 
    DISTANCE JOIN 
        p2 
    ON 
        POINT(p2.lng, p2.lat) IN CIRCLERANGE(POINT(p1.lng, p1.lat), {0})""".format(epsilon)
pairs = sqlContext.sql(sql).rdd.filter(lambda pair: pair[0] < pair[3]).map(calculateDisks)
ndisks = pairs.count()
t2 = round(time.time() - t1,3)
print("PBFE-SQL-POSTFILTER,{0},{1},{2},{3}".format(float(epsilon),tag,2 * ndisks,t2))



